src/phoenix/db/bulk_inserter.py

```diff
@@ -0,0 +1,177 @@
+import asyncio
+import logging
+from itertools import islice
+from time import time
+from typing import Any, AsyncContextManager, Callable, Iterable, List, Optional, Tuple, cast
+
+from openinference.semconv.trace import SpanAttributes
+from sqlalchemy import func, insert, select, update
+from sqlalchemy.ext.asyncio import AsyncSession
+
+from phoenix.db import models
+from phoenix.trace.schemas import Span, SpanStatusCode
+
+logger = logging.getLogger(__name__)
+
+
+class SpansBulkInserter:
+    def __init__(
+        self,
+        db: Callable[[], AsyncContextManager[AsyncSession]],
+        initial_batch: Optional[Iterable[Tuple[Span, str]]] = None,
+        sleep_seconds: float = 0.5,
+        max_num_spans_per_transaction: int = 100,
+    ) -> None:
+        self._db = db
+        self._running = False
+        self._sleep_seconds = sleep_seconds
+        self._max_num_spans_per_transaction = max_num_spans_per_transaction
+        self._batch: List[Tuple[Span, str]] = [] if initial_batch is None else list(initial_batch)
+        self._task: Optional[asyncio.Task[None]] = None
+
+    async def __aenter__(self) -> Callable[[Span, str], None]:
+        self._running = True
+        self._task = asyncio.create_task(self._insert_spans())
+        return self._queue_span
+
+    async def __aexit__(self, *args: Any) -> None:
+        self._running = False
```

Comment:
Set `self._task` to None?

Reply:
i'll let the garbage collect do it later. it's harmless either way